runtime/parachains/src/scheduler_parathreads/mod.rs

```diff
@@ -0,0 +1,459 @@
+//! - Parachains that have a candidate pending availability in this fork of the chain
+//!   should not be assigned.
+//! - Validator assignments should not be gameable. Malicious cartels should not be able to
+//!   manipulate the scheduler to assign themselves as desired.
+//! - High or close to optimal throughput of parachains and parathreads. Work among validator groups should be balanced.
+//!
+//! The Scheduler manages resource allocation using the concept of "Availability Cores".
+//! There will be one availability core for each parachain, and a fixed number of cores
+//! used for multiplexing parathreads. Validators will be partitioned into groups, with the same
+//! number of groups as availability cores. Validator groups will be assigned to different availability cores
+//! over time.
+
+use frame_support::pallet_prelude::*;
+use primitives::{
+	CoreIndex, CoreOccupied, GroupIndex, Id as ParaId, ParathreadClaim, ParathreadEntry,
+	ScheduledCore,
+};
+use scale_info::TypeInfo;
+use sp_runtime::Saturating;
+use sp_std::{collections::btree_map::BTreeMap, prelude::*};
+
+use crate::{
+	configuration,
+	initializer::SessionChangeNotification,
+	paras,
+	scheduler_common::{AssignmentKind, CoreAssigner, CoreAssignment, FreedReason},
+};
+
+pub use pallet::*;
+
+//#[cfg(test)]
+//mod tests;
+
+/// A queued parathread entry, pre-assigned to a core.
+#[derive(Encode, Decode, TypeInfo)]
+#[cfg_attr(test, derive(PartialEq, Debug))]
+pub struct QueuedParathread {
+	pub claim: ParathreadEntry,
+	pub core_offset: u32,
+}
+
+/// The queue of all parathread claims.
+#[derive(Encode, Decode, TypeInfo)]
+#[cfg_attr(test, derive(PartialEq, Debug))]
+pub struct ParathreadClaimQueue {
+	pub queue: Vec<QueuedParathread>,
+	// this value is between 0 and config.parathread_cores
+	pub next_core_offset: u32,
+}
+
+impl ParathreadClaimQueue {
+	/// Queue a parathread entry to be processed.
+	///
+	/// Provide the entry and the number of parathread cores, which must be greater than 0.
+	pub fn enqueue_entry(&mut self, entry: ParathreadEntry, n_parathread_cores: u32) {
+		let core_offset = self.next_core_offset;
+		self.next_core_offset = (self.next_core_offset + 1) % n_parathread_cores;
+
+		self.queue.push(QueuedParathread { claim: entry, core_offset })
+	}
+
+	/// Take next queued entry with given core offset, if any.
+	fn take_next_on_core(&mut self, core_offset: u32) -> Option<ParathreadEntry> {
+		let pos = self.queue.iter().position(|queued| queued.core_offset == core_offset);
+		pos.map(|i| self.queue.remove(i).claim)
+	}
+
+	/// Get the next queued entry with given core offset, if any.
+	pub fn get_next_on_core(&self, core_offset: u32) -> Option<&ParathreadEntry> {
+		let pos = self.queue.iter().position(|queued| queued.core_offset == core_offset);
+		pos.map(|i| &self.queue[i].claim)
+	}
+}
+
+impl Default for ParathreadClaimQueue {
+	fn default() -> Self {
+		Self { queue: vec![], next_core_offset: 0 }
+	}
+}
+
+pub struct ParathreadsScheduler;
```

Comment:
You can also implement `CoreAssigner<T>` directly for the pallet struct and avoid the indirection of the pallet submodule.

This maps nicely into `construct_runtime!` usages in runtime declarations, where the pallet is given a concrete name (e.g. ParathreadsScheduler). The generated `ParathreadsScheduler = scheduler_parathreads::Pallet<Runtime>` type alias can be used directly without having to import another type.

Reply:
That would be more idiomatic in FRAME; 

e.g. https://github.com/paritytech/substrate/blob/ea002409d10af3fb7a402b6948303233c6f46549/frame/election-provider-multi-phase/src/lib.rs#L1640
https://github.com/paritytech/substrate/blob/b3d9f3c57e7f799d49442e3456614431f2d0e51a/frame/ranked-collective/src/lib.rs#L235
https://github.com/paritytech/substrate/blob/ea002409d10af3fb7a402b6948303233c6f46549/frame/referenda/src/lib.rs#L639